libraries/Microsoft.Bot.Builder.Core/Middleware/StateMiddleware.cs

```diff
@@ -0,0 +1,174 @@
+            if (!_settings.WriteBeforeSend)
+            {
+                await Write(context).ConfigureAwait(false);
+            }
+        }
+
+        protected virtual async Task<StoreItems> Read(IBotContext context)
+        {
+            var key = this.GetStorageKey(context);
+            var keys = new List<String>();
+            keys.Add(key);
+            var items = await _storage.Read(keys.ToArray());
+            var state = items.Get<StateT>(key);
+            if (state == null)
+                state = new StateT();
+            context.Set(GetPropertyName(), state);
+            return items;
+        }
+
+        protected virtual async Task Write(IBotContext context, StoreItems changes = null)
+        {
+            if (changes == null)
+                changes = new StoreItems();
+
+            var state = context.Get<StateT>(GetPropertyName());
+            if (state == null)
+                state = new StateT();
+            var key = GetStorageKey(context);
+            changes[key] = state;
+
+            if (this._settings.LastWriterWins)
+            {
+                foreach (var item in changes)
+                {
+                    ((StoreItem)changes[item.Key]).eTag = "*";
+                }
+            }
+
+            await _storage.Write(changes).ConfigureAwait(false);
+        }
+
+    }
+
+    public class ConversationStateMiddleware<StateT> : StateMiddleware<StateT>
+        where StateT : IStoreItem, new()
+    {
+
+        public ConversationStateMiddleware(IStorage storage, StateSettings settings = null) :
+            base(storage, settings)
+        {
+        }
+
+        public static readonly string PropertyName = $"{typeof(ConversationStateMiddleware<StateT>).Namespace}.{typeof(ConversationStateMiddleware<StateT>).Name}";
+
+        public override string GetPropertyName()
+        {
+            return PropertyName;
+        }
+
+        public override string GetStorageKey(IBotContext context)
+        {
+            var conversation = context.ConversationReference;
+            return $"conversation/{conversation.ChannelId}/{conversation.Conversation.Id}";
+        }
+    }
+
+    public class UserStateMiddleware<StateT> : StateMiddleware<StateT>
+        where StateT : IStoreItem, new()
+    {
+
+        public UserStateMiddleware(IStorage storage, StateSettings settings = null) :
+            base(storage, settings)
+        {
+        }
+
+        public static readonly string PropertyName = $"{typeof(UserStateMiddleware<StateT>).Namespace}.{typeof(UserStateMiddleware<StateT>).Name}";
+
+        public override string GetPropertyName()
+        {
+            return PropertyName;
+        }
+
+        public override string GetStorageKey(IBotContext context)
+        {
+            var conversation = context.ConversationReference;
+            return $"user/{conversation.ChannelId}/{conversation.User.Id}";
+        }
+    }
+
+    public static class StateContextExtensions
```

Comment:
>StateContextExtensions [](start = 24, length = 22)

This is what Bill keeps pushing back on. He doesn't think we should EVER add extension methods because we can't in Java and what if some other plugin wanted to add something called GetConversationState() that worked differently. You'd have a collision..

I personally hate the lowest common denominator approach of that and you already know what I think of the collision argument but others need to push back on that.